velox/expression/Expr.cpp

```diff
@@ -567,96 +563,82 @@ void Expr::eval(
-        const_cast<const SelectivityVector*>(newFinalSelection),
-        true /*checkCondition*/,
-        true /*override*/);
-
-    evalEncodings(*missingRows, context, sharedSubexprValues_);
+    *sharedSubexprRows_ = rows;
+    if (context.errors()) {
+      // Clear the rows which failed to compute.
+      context.deselectErrors(*sharedSubexprRows_);
+      if (!sharedSubexprRows_->hasSelections()) {
+        // Do not store a reference to 'result' if we cannot use any rows from
+        // it.
+        return;
+      }
+    }
 
-    // Clear the rows which failed to compute.
-    context.deselectErrors(*sharedSubexprRows_);
+    sharedSubexprValues_ = result;
+    return;
   }
 
-  context.moveOrCopyResult(sharedSubexprValues_, rows, result);
-  return true;
-}
-
-void Expr::checkUpdateSharedSubexprValues(
-    const SelectivityVector& rows,
-    EvalCtx& context,
-    const VectorPtr& result) {
-  if (!isMultiplyReferenced_ || sharedSubexprValues_ ||
-      context.wrapEncoding() != VectorEncoding::Simple::FLAT) {
+  if (rows.isSubset(*sharedSubexprRows_)) {
+    // We have results for all requested rows. No need to compute anything.
+    context.moveOrCopyResult(sharedSubexprValues_, rows, result);
     return;
   }
 
-  if (!sharedSubexprRows_) {
-    sharedSubexprRows_ = context.execCtx()->getSelectivityVector(rows.size());
-  }
+  // We are missing results for some or all of the requested rows. Need to
+  // compute these and save for future use.
 
-  *sharedSubexprRows_ = rows;
-  if (context.errors()) {
-    // Clear the rows which failed to compute.
-    context.deselectErrors(*sharedSubexprRows_);
-    if (!sharedSubexprRows_->hasSelections()) {
-      // Do not store a reference to 'result' if we cannot use any rows from it.
-      return;
-    }
+  // Identify a subset of rows that need to be computed: rows -
+  // sharedSubexprRows_.
+  LocalSelectivityVector missingRowsHolder(context, rows);
+  auto missingRows = missingRowsHolder.get();
+  missingRows->deselect(*sharedSubexprRows_);
+  VELOX_DCHECK(missingRows->hasSelections());
+
+  // Fix finalSelection to avoid losing values outside missingRows.
+  // Final selection of rows need to include sharedSubexprRows_, missingRows and
+  // current final selection of rows if set.
+  LocalSelectivityVector newFinalSelectionHolder(context, *sharedSubexprRows_);
+  auto newFinalSelection = newFinalSelectionHolder.get();
+  newFinalSelection->select(*missingRows);
+  if (!context.isFinalSelection()) {
+    newFinalSelection->select(*context.finalSelection());
   }
 
-  sharedSubexprValues_ = result;
+  ScopedFinalSelectionSetter setter(
+      context, newFinalSelection, true /*checkCondition*/, true /*override*/);
+
+  evalEncodings(*missingRows, context, sharedSubexprValues_);
+
+  // Clear the rows which failed to compute.
+  context.deselectErrors(*missingRows);
+
+  sharedSubexprRows_->select(*missingRows);
+  context.moveOrCopyResult(sharedSubexprValues_, *missingRows, result);
```

Comment:
The "missingRows" over here only copies missing rows and we lose out on the rows that were shared. instead it needs the full set of "rows" minus the rows that errored out which we remove in L638.
See #3958

Reply:
@bikramSingh91 Thank you for discovering this bug. It needs to be `context.moveOrCopyResult(sharedSubexprValues_, rows, result);` I think.